sale_start_end_dates/tests/test_sale_start_end_dates.py

```diff
@@ -0,0 +1,122 @@
+                {
+                    "default_start_date": self.default_end_date,
+                    "default_end_date": self.default_start_date,
+                }
+            )
+        self.so.default_start_date_change()
+
+    def test_default_end_date_change(self):
+        with self.assertRaises(ValidationError):
+            self.so.write(
+                {
+                    "default_start_date": self.default_end_date,
+                    "default_end_date": self.default_start_date,
+                }
+            )
+        self.so.default_end_date_change()
+
+    def test_start_end_dates_product_id_change(self):
+        if self.so.default_end_date and self.so.default_end_date:
+            self.so.order_line.start_end_dates_product_id_change()
+            self.so.order_line.start_date_change()
+            self.so.order_line.end_date_change()
+        self.product_id.must_have_dates = False
+        self.so.order_line.start_end_dates_product_id_change()
+
+    def test_start_end_dates_product_id(self):
+        self.product_id.must_have_dates = False
+        self.so.default_start_date = self.so.default_end_date = False
+        self.so.order_line.start_end_dates_product_id_change()
+
+    def test_end_date_change(self):
+        self.product_id.must_have_dates = False
+        self.so.order_line.write(
+            {"start_date": self.default_end_date, "end_date": self.default_start_date}
+        )
+        self.so.order_line.end_date_change()
+
+    def test_start_date_change(self):
+        self.product_id.must_have_dates = False
+        self.so.order_line.write(
+            {"start_date": self.default_end_date, "end_date": self.default_start_date}
+        )
+        self.so.order_line.start_date_change()
+
+    def test_constrains_end_dates(self):
+        with self.assertRaises(ValidationError):
+            self.so.order_line.end_date = False
+
+    def test_constrains_start_date(self):
+        with self.assertRaises(ValidationError):
+            self.so.order_line.start_date = False
+
+    def test_constrains_greater_st_date(self):
+        with self.assertRaises(ValidationError):
+            self.so.order_line.write(
+                {
+                    "start_date": self.default_end_date,
+                    "end_date": self.default_start_date,
+                }
+            )
+
+    def test_compute_number_of_days(self):
+        self.assertEqual(self.so.order_line[0].number_of_days, 10)
+
+    def test_inverse_number_of_days(self):
+        self.so.order_line[0].number_of_days = 1
+        self.assertEqual(
+            self.so.order_line[0].start_date, self.so.order_line[0].end_date
+        )
+        self.so.order_line[0].number_of_days = -1
+        self.assertEqual(self.so.order_line[0].number_of_days, 1)
+
+    def test_prepare_invoice_line(self):
+        invoice_line_vals = self.so.order_line._prepare_invoice_line()
+        self.assertEqual(invoice_line_vals["product_id"], self.product_id.id)
```

Comment:
we can add two more lines here to check `start_date` and `end_date` on invoice line

```
self.assertEqual(invoice_line_vals["start_date"], self.so.order_line[0].start_date)
self.assertEqual(invoice_line_vals["end_date"], self.so.order_line[0].end_date)
```

Reply:
@chandni299 i added these lines